Tracer interpolation to field

gassmoeller opened the pull request:
This is a preliminary implementation of how to consistently interpolate from particles to compositional fields. It contains bugs and is not in any shape for review, I just open the PR so that interested early adopters can look at it.

gassmoeller commented:
This should be ready for a review. It still has the drawback of requiring initial conditions for the particle field, but I am working on that in a separate pull request (allowing lists of initial conditions for temperature and composition, and consequently not prescribing ic's for particle fields)

tjhei commented on include/aspect/introspection.h:
```diff
@@ -293,6 +293,13 @@ namespace aspect
        */
       IndexSets index_sets;
 
+      enum AdvectionMethod
```
In many places we wrap the enum in a struct an call the enum itself ``kind`` to scope the enum values.

tjhei commented on include/aspect/introspection.h:
```diff
@@ -293,6 +293,13 @@ namespace aspect
+        field = 0,
+        particles = 1
+      };
+      std::map<std::string,AdvectionMethod> advection_methods;
```
Is there a reason you go with a lookup by name instead of using the index (and put it in a std::vector)? The advantage would be that you don't have to worry about the case that you can not find the entries. Also please add comments to this variable.

tjhei commented on include/aspect/parameters.h:
```diff
@@ -292,6 +292,15 @@ namespace aspect
+     * where component is of the format "[1][2][3]" and the particle property is
+     * mapped to one of the particle property plugins.
+     */
+    std::map<std::string, std::pair<std::string,unsigned int> > mapped_particle_properties;
```
again, why do you go by name here?

tjhei commented on include/aspect/parameters.h:
```diff
@@ -292,6 +292,15 @@ namespace aspect
+
+    /**
+     * Map from compositional field name to a pair "particle property", "component",
+     * where component is of the format "[1][2][3]" and the particle property is
```
I don't understand ``format "[1][2][3]"`` because you store it in an int? Are you trying to say that the component is a value between 1 and dim?

tjhei commented on include/aspect/particle/interpolator/interface.h:
```diff
@@ -63,18 +63,35 @@ namespace aspect
+           * a cell iterator to the cell that contains the @p positions. Often the
+           * cell containing the positions is known and this function can be implemented
+           * more efficiently than the general one above. It has a default implementation
+           * that simply calls the function with two arguments.
```
Is there a reason you don't create a single function that gets a cell ptr that can be NULL?

bangerth replied:
Instead of a pointer to a cell iterator, just use a cell iterator and use `cell_iterator()` to mark an invalid iterator.

tjhei commented on include/aspect/simulator.h:
```diff
@@ -242,6 +242,13 @@ namespace aspect
 
         /**
+         * Return whether this object refers to a field discretized by
+         * discontinuous finite elements.
```
I thought it returns particle or field?

tjhei commented on source/simulator/initial_conditions.cc:
```diff
@@ -174,6 +174,104 @@ namespace aspect
+    // pressure element is non-interpolating (we get an exception
+    // even though we are, strictly speaking, not interested in
+    // interpolating the pressure; but, as mentioned, there is no way
+    // to tell VectorTools::interpolate that)
```
yes, :-(

bangerth replied:
How do we do this for the initial temperature field then? There must be code there that we could steal for here.

gassmoeller replied:
This is a copy of the code for the initial temperature :smile:

bangerth replied:
Hm. I wonder if we could somehow extract that code and put it into a common function, rather than duplicate code. Alternatively, we may also extend the deal.II function and call that when available -- in that case, the code duplication would eventually go away as we require a newer deal.II version.

tjhei commented on source/simulator/parameters.cc:
```diff
@@ -723,6 +723,33 @@ namespace aspect
                          Patterns::List(Patterns::Anything()),
                          "A user-defined name for each of the compositional fields requested.");
+      prm.declare_entry ("Advection method of fields", "",
+                         Patterns::Map (Patterns::Anything(),
```
why not use a list of length n_compositional_fields with pattern(field|particles)?

tjhei commented:
I would propose to rename the new property to be something like "compositional field type" so we can later add other types like "level set", "percentage", etc.. Thoughts?

gassmoeller commented:
Second round of review please. I am sure there will be more :smile:

bangerth commented on include/aspect/introspection.h:
```diff
@@ -294,6 +294,28 @@ namespace aspect
+        {
+          continuous_fem_field = 0,
+          discontinuous_fem_field = 1,
+          particles = 2
```
Can you document these three choices? I also don't think you actually need the initializers -- the symbolic names should be enough.

While there, rename `kind` -> `Kind`.

bangerth replied:
Having read through the patch, the only place where you seem to make a distinction between the continuous and discontinuous case is in the place where you call the solver and, possibly, the limiter. But this is already superseded by #1109. Would it make sense to merge the first two enum entries and just call it `fem_field` or similar?

gassmoeller replied:
Only one `fem_field` was my first implementation, but with the distinction we can allow for different fields to have different finite element types (which was on our long-term to-do list anyway). This enum could also then supersede the `use_discontinuous_composition_discretization` variable that does not make sense with other advection methods like particles. This is not fully implemented in this PR, because I wanted to discuss it first.

bangerth replied:
I see. I just want to avoid us using different methods to achieve the same goal in different places. Partial transitions are one of the biggest problems in software engineering.

I'm ok with this change if you find a way to remove the existing boolean flags from the `Simulator` class and somehow parse the existing parameters into the flags encoded by the current parameters. Or if you wanted to get rid of these two parameters altogether, that would be ok as well.

gassmoeller replied:
I looked a bit into the issue and there are just too many places that are affected by this to include it in this PR. The option to have different finite elements for different compositional fields (including the one that are advected on particles) will have to wait for another PR.

bangerth commented on include/aspect/introspection.h:
```diff
@@ -294,6 +294,28 @@ namespace aspect
+       * compositional fields. This type is used to determine how
+       * to advect a compositional field.
+       */
+      struct FieldMethod
```
Is this going to be used for anything other than the compositional fields? `FieldMethod` is not very descriptive. Maybe call it `CompositionalFieldComputation`? I don't particularly care for that name either, so if you can come up with something better, go with that!

gassmoeller replied:
I was playing with the thought of adding the option that the temperature field might be solved differently as well (as far as I know some codes in Europe do that), but currently it really only applies to compositional fields. The complication in finding a good name comes from the definition of "field", i.e. does it refer to one "composition" or does it refer to a "field method". Would you agree on CompositionalFieldMethod? I also thought about `AdvectionMethod`, but I really do not feel like any name is as descriptive as necessary.

bangerth replied:
Yes, I struggled with naming as well. I don't think I have a better suggestion -- `CompositionalFieldMethod` or `AdvectedFieldMethod` are both ok with me, depending on whether you want to include the temperature or not. I guess including the temperature makes sense if you want to use the thing here to discriminate between CG and DG methods.

I think all I wanted to say is that "FieldMethod" is just a bit too unspecific ;-)

bangerth commented on include/aspect/parameters.h:
```diff
@@ -293,6 +293,24 @@ namespace aspect
+     * A vector that contains the field method for every compositional
+     * field. Consequently the vector has n_compositional_fields entries.
+     */
+    std::vector<std::string> compositional_field_methods;
```
I think it would be better to make this a vector of enums, i.e., do the translation at parse time, rather than every time you have to look it up.

bangerth commented on include/aspect/parameters.h:
```diff
@@ -293,6 +293,24 @@ namespace aspect
+     * the component and it is of the format "[0][1][2]". In case no component
+     * is specified it defaults to 0.
+     */
+    std::map<std::string, std::pair<std::string,unsigned int> > mapped_particle_properties;
```
Could you make the keys of this map an index within the compositional fields?

bangerth replied:
You'll of course have to do the translation somewhere, but I think you can do that at the end of `parse_parameters()`, can't you?

bangerth commented on include/aspect/particle/interpolator/cell_average.h:
```diff
@@ -0,0 +1,55 @@
+          /**
+           * @copydoc aspect::Particle::Interpolator::Interface::properties_at_points()
+           */
+          virtual
```
if you say nothing at all, then it will automatically copy the documentation from the corresponding function in the base class.

bangerth commented on include/aspect/particle/interpolator/cell_average.h:
```diff
@@ -0,0 +1,55 @@
+    namespace Interpolator
+    {
+      /**
+       * Return the averaged properties of all tracers of the given cell.
```
of -> on

bangerth commented on include/aspect/simulator.h:
```diff
@@ -245,6 +245,13 @@ namespace aspect
+         * (i.e. 'field', 'particles').
+         */
+        typename Introspection<dim>::FieldMethod::kind
+        method (const Introspection<dim> &introspection) const;
```
can we come up with a better name for this function?

bangerth commented on source/particle/interpolator/cell_average.cc:
```diff
@@ -0,0 +1,108 @@
+       * Return the cell-wise averaged properties of all tracers of the cell containing the
+       * given positions.
+       */
+      template <int dim>
```
no need for the two asterisks since this is not a doxygen comment

bangerth commented on source/particle/interpolator/cell_average.cc:
```diff
@@ -0,0 +1,108 @@
+            // We can not simply use one of the points as input for find_active_cell_around_point
+            // because for vertices of mesh cells we might end up getting ghost_cells as return value
+            // instead of the local active cell. So make sure we are well in the inside of a cell.
+            Point<dim> approximated_cell_midpoint = positions[0];
```
const

bangerth replied:
oh, forget about it

bangerth replied:
but different question: what do you do if there are no points? I think we should at least assert that there are.

bangerth commented on source/particle/interpolator/cell_average.cc:
```diff
@@ -0,0 +1,108 @@
+                for (unsigned int i = 1; i<positions.size()-1; ++i)
+                  direction_to_center += positions[i] - positions[0];
+                direction_to_center /= positions.size() - 1;
+                approximated_cell_midpoint += direction_to_center;
```
is this algorithm a complicated way of computing (or approximating) the mean value of all of the positions? If so, why not just
```
  approximated_c_m = std::accumulate (positions.begin(), positions.end(), Point<dim>()) / positions.size();
```
?

bangerth commented on source/particle/interpolator/cell_average.cc:
```diff
@@ -0,0 +1,108 @@
+        for (typename std::multimap<types::LevelInd, Particle<dim> >::const_iterator particle = particle_range.first;
+             particle != particle_range.second; ++particle)
+          {
+            const std::vector<double> particle_properties = particle->second.get_properties();
```
does `get_properties()` return a vector, or a reference to a vector? If the latter, make this variable a reference

bangerth commented on source/particle/interpolator/cell_average.cc:
```diff
@@ -0,0 +1,108 @@
+            const std::vector<double> particle_properties = particle->second.get_properties();
+
+            for (unsigned int i = 0; i < n_properties; ++i)
+              cell_properties[i] += particle_properties[i] / n_particles;
```
This is inefficient -- just do the accumulation first, and then divide by `n_particles` once at the end.

bangerth commented on source/particle/interpolator/cell_average.cc:
```diff
@@ -0,0 +1,108 @@
+
+        const std::vector<std::vector<double> > properties(positions.size(),cell_properties);
+
+        return properties;
```
I think it might be just as easy to not create a named variable but just return the vector right away. This won't make much of a difference in C++11 but should also be faster on older compilers.

bangerth commented on source/simulator/core.cc:
```diff
@@ -1786,11 +1786,29 @@ namespace aspect
+                    assemble_advection_system (adv_field);
+                    solve_advection(adv_field);
+                    if (parameters.use_limiter_for_discontinuous_composition_solution)
+                      apply_limiter_to_dg_solutions(adv_field);
```
This conflicts with the changes made in #1109. Can you rebase?

bangerth commented on source/simulator/parameters.cc:
```diff
@@ -739,6 +739,28 @@ namespace aspect
                          Patterns::List(Patterns::Anything()),
                          "A user-defined name for each of the compositional fields requested.");
+      prm.declare_entry ("Compositional field methods", "",
+                         Patterns::List (Patterns::Selection("continuous field|particles")),
```
This list of choices does not actually allow for a discontinuous FEM field :-)

bangerth commented on source/simulator/parameters.cc:
```diff
@@ -739,6 +739,28 @@ namespace aspect
                          "A user-defined name for each of the compositional fields requested.");
+      prm.declare_entry ("Compositional field methods", "",
+                         Patterns::List (Patterns::Selection("continuous field|particles")),
+                         "A comma separated list denoting the solution method of each "
```
of -> for

bangerth commented on source/simulator/parameters.cc:
```diff
@@ -739,6 +739,28 @@ namespace aspect
+                         "compositional field. Each entry of the list must be "
+                         "one of the currently implemented field types: "
+                         "``continuous_fem_field'', ``discontinuous_fem_field'', "
+                         "or ``particles''.");
```
I think it would be useful to document what each of these choices actually *does*.

The spelling used here is different from the one given in the `Patterns::Selection` above. They should match.

bangerth commented on source/simulator/parameters.cc:
```diff
@@ -739,6 +739,28 @@ namespace aspect
+      prm.declare_entry ("Mapped particle properties", "",
+                         Patterns::Map (Patterns::Anything(),
+                                        Patterns::Anything()),
+                         "A comma separated list denoting the particle properties "
```
If it is a comma separated list, shouldn't the pattern then be a `Patterns::List(Patterns::Map(...), ',')`?

bangerth commented on source/simulator/parameters.cc:
```diff
@@ -739,6 +739,28 @@ namespace aspect
+                         "selected particle properties. Component is a component index of "
+                         "the particle property that is 0 by default, and only needs to be "
+                         "set if a different component of the particle property should be "
+                         "mapped (e.g. the y-component of the velocity at the particle positions).");
```
-> `$y$`

I think what the documentation is lacking is a half-sentence that indicates that particle property names may imply multiple properties (e.g., the particle position implies a *vector* of properties), and that the component selector picks one of those if there are in fact multiple.

bangerth commented on source/simulator/parameters.cc:
```diff
@@ -1124,6 +1146,124 @@ namespace aspect
+      // If no method is specified set the default, which is solve every composition
+      // by a continuous field method
+      if (compositional_field_methods.size() == 0)
+        compositional_field_methods = std::vector<std::string> (n_compositional_fields,"continuous field");
```
ah, so this would override the current way of selecting between continuous and discontinuous field by looking at the boolean flag that makes the decision for *all* compositional fields. Is your intent to be incompatible? It would make sense, since your approach allows making this decision field by field, rather than for *all* compositional fields at once.

bangerth commented on source/simulator/parameters.cc:
```diff
@@ -1124,6 +1146,124 @@ namespace aspect
+                   || (x_mapped_particle_properties.size() == 0),
+                   ExcMessage ("The list of names for the mapped particle "
+                               "property fields needs to either be empty or have a length equal to "
+                               "the number of compositional fields that are advected on particles."));
```
maybe say "...that are interpolated from particle properties"? I think that's what happening, right?

bangerth commented on source/simulator/parameters.cc:
```diff
@@ -1124,6 +1146,124 @@ namespace aspect
+                                   "<name of field> : <particle property> [component]', "
+                                   "but there does not appear to be a colon in the entry <"
+                                   + *p
+                                   + ">."));
```
nice error message!

bangerth commented on source/simulator/parameters.cc:
```diff
@@ -1124,6 +1146,124 @@ namespace aspect
+                                   "> appears in the list of mapped particle properties, but"
+                                   "there is no field with this name."));
+
+          const unsigned int compositional_field_index = std::distance(names_of_compositional_fields.begin(),field_name_iterator);
```
break this line

bangerth commented on source/simulator/parameters.cc:
```diff
@@ -1124,6 +1146,124 @@ namespace aspect
+                                  names_of_compositional_fields.end(), key) == 1,
+                       ExcMessage ("Name of field <" + key +
+                                   "> appears more than once in the list of "
+                                   "mapped particle properties."));
```
maybe name the parameter in which this happens to make it easier to find the place in the parameter file where this happens?

bangerth commented on source/simulator/parameters.cc:
```diff
@@ -1124,6 +1146,124 @@ namespace aspect
+          if ((value_and_comp.size()>0) &&
+              ((value_and_comp[value_and_comp.size()-1] >= '0')
+               &&
+               (value_and_comp[value_and_comp.size()-1] <= '9')))
```
I had assumed from the description in the `declare_parameter` call that one would append the component selector in square brackets, in the spirit of an indexing operation. I realize now that you did not mean the square brackets that way, but as optional. Either way, it might be easier to actually *use* square brackets to select a component, because it would make parsing so much simpler. Want to consider?

bangerth commented on tests/discontinuous_composition_bound_preserving_limiter/screen-output:
```diff
@@ -9,8 +9,8 @@ Number of degrees of freedom: 8,164 (2,178+289+1,089+2,304+2,304)
-Number of active cells: 157 (on 6 levels)
-Number of degrees of freedom: 5,244 (1,482+195+741+1,413+1,413)
+Number of active cells: 151 (on 6 levels)
+Number of degrees of freedom: 5,035 (1,420+187+710+1,359+1,359)
```
where does this difference come from? Shouldn't this patch make no difference to existing functionality?

egpuckett commented:
What is the current status of this PR? @hlokavarapu and I are trying to benchmark the accuracy of the active tracers with the SolKz, SolCx, and SolVI  (inclusion) benchmarks as in Thielmann M., May D.A., Kaus (2014). I'm not clear as to whether I should start with the ASPECT master branch or work from this branch.

bangerth commented:
Good question. @gassmoeller ?

gassmoeller commented:
I think I addressed all comments. I will try to make the tester happy and then this should be ready to merge. Some details may need some fine-tuning, but the parameter file syntax should be ready, and since many follow-up work depends on this PR, we should go ahead and fix other things later. Sorry for being slow on this.

bangerth commented:
OK. I'm still curious why there are differences in the output of existing tests.

Let me know once you tink it's ready.

gassmoeller commented:
Lets see if this is still the case when the tester is finished. I have the feeling that was a problem during rebasing and a conflict with another pull request.

gassmoeller commented:
I have rebased, updated all test results and added an entry to changes.h. After rebasing the only old test that changes is the one where particles are added during the computation, and that is because of the change in the default behavior for the particle property initialization for the 'initial position' property. The new default is much more reasonable though, so the changed test result actually represents something between an improvement and a bug fix. So this PR should be ready to merge as soon as the tester is finished.

bangerth commented:
OK to merge from my perspective.
